# netsurf: utils
# 14 files, 304 changed lines
# MacSurf fix ids: fixes1063 fixes161f fixes195 fixes907 fixes908 fixes914 fixes927
#
# Regenerate: tools/regen-patches.sh && tools/split-patches.py
#
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/config.h browser/netsurf/utils/config.h
--- upstream/netsurf/utils/config.h	2026-09-20 23:29:00.183307243 +0100
+++ browser/netsurf/utils/config.h	2026-09-20 23:19:21.051021423 +0100
@@ -41,13 +41,24 @@
      defined(__amigaos4__) || \
      defined(__HAIKU__) || \
      (defined(_POSIX_C_SOURCE) && ((_POSIX_C_SOURCE - 0) >= 200809L)) && \
-     !defined(__riscos__))
+     !defined(__riscos__) && \
+     !defined(__MACOS9__))
 #define HAVE_STRNDUP
 #else
 #undef HAVE_STRNDUP
 char *strndup(const char *s, size_t n);
 #endif
 
+#if defined(__MACOS9__)
+#ifndef PATH_MAX
+#define PATH_MAX 1024
+#endif
+char *strdup(const char *s);
+int ftruncate(int fd, long long length);
+int strcasecmp(const char *s1, const char *s2);
+int strncasecmp(const char *s1, const char *s2, size_t n);
+#endif
+
 #if ((defined(_GNU_SOURCE) ||			\
       defined(NS_NEW_GLIBC) || \
       defined(__APPLE__) ||			\
@@ -70,7 +81,8 @@
      defined(__BEOS__) ||     \
      defined(__amigaos4__) || \
      defined(__AMIGA__) ||    \
-     defined(__MINT__))
+     defined(__MINT__) ||     \
+     defined(__MACOS9__))
 #undef HAVE_STRPTIME
 #undef HAVE_STRFTIME
 #else
@@ -78,6 +90,11 @@
 #define HAVE_STRFTIME
 #endif
 
+/* Mac OS 9 provides strftime via mac_time.c shim */
+#if defined(__MACOS9__)
+#define HAVE_STRFTIME 1
+#endif
+
 /* For some reason, UnixLib defines this unconditionally. Assume we're using
  *  UnixLib if building for RISC OS.
  */
@@ -102,29 +119,30 @@
 
 #define HAVE_SYS_SELECT
 #define HAVE_POSIX_INET_HEADERS
-#if (defined(_WIN32))
+#if (defined(_WIN32) || defined(__MACOS9__))
 #undef HAVE_SYS_SELECT
 #undef HAVE_POSIX_INET_HEADERS
 #endif
 
 #define HAVE_INETATON
 #if (defined(_WIN32) || \
-     defined(__serenity__))
+     defined(__serenity__) || \
+     defined(__MACOS9__))
 #undef HAVE_INETATON
 #endif
 
 #define HAVE_INETPTON
-#if (defined(_WIN32))
+#if (defined(_WIN32) || defined(__MACOS9__))
 #undef HAVE_INETPTON
 #endif
 
 #define HAVE_UTSNAME
-#if (defined(_WIN32))
+#if (defined(_WIN32) || defined(__MACOS9__))
 #undef HAVE_UTSNAME
 #endif
 
 #define HAVE_REALPATH
-#if (defined(_WIN32))
+#if (defined(_WIN32) || defined(__MACOS9__))
 #undef HAVE_REALPATH
 char *realpath(const char *path, char *resolved_path);
 #endif
@@ -135,7 +153,7 @@
 #endif
 
 #define HAVE_SIGPIPE
-#if (defined(_WIN32))
+#if (defined(_WIN32) || defined(__MACOS9__))
 #undef HAVE_SIGPIPE
 #endif
 
@@ -145,27 +163,28 @@
 #endif
 
 #define HAVE_MMAP
-#if (defined(_WIN32) || defined(__riscos__) || defined(__HAIKU__) || defined(__BEOS__) || defined(__amigaos4__) || defined(__AMIGA__) || defined(__MINT__))
+#if (defined(_WIN32) || defined(__riscos__) || defined(__HAIKU__) || defined(__BEOS__) || defined(__amigaos4__) || defined(__AMIGA__) || defined(__MINT__) || defined(__MACOS9__))
 #undef HAVE_MMAP
 #endif
 
 #define HAVE_SCANDIR
 #if (defined(_WIN32) ||				\
-     defined(__serenity__))
+     defined(__serenity__) ||			\
+     defined(__MACOS9__))
 #undef HAVE_SCANDIR
 #endif
 
 #define HAVE_DIRFD
 #define HAVE_UNLINKAT
 #define HAVE_FSTATAT
-#if (defined(_WIN32) || defined(__riscos__) || defined(__HAIKU__) || defined(__BEOS__) || defined(__amigaos4__) || defined(__AMIGA__) || defined(__MINT__))
+#if (defined(_WIN32) || defined(__riscos__) || defined(__HAIKU__) || defined(__BEOS__) || defined(__amigaos4__) || defined(__AMIGA__) || defined(__MINT__) || defined(__MACOS9__))
 #undef HAVE_DIRFD
 #undef HAVE_UNLINKAT
 #undef HAVE_FSTATAT
 #endif
 
 #define HAVE_REGEX
-#if (defined(__serenity__))
+#if (defined(__serenity__) || defined(__MACOS9__))
 #undef HAVE_REGEX
 #endif
 
@@ -185,6 +204,9 @@
 #if defined(riscos)
     /* Theme auto-install */
     #define WITH_THEME_INSTALL
+#elif defined(__MACOS9__)
+    /* Mac OS 9 / Carbon - cooperative multitasking, no POSIX layer */
+    #define NO_IPV6 1
 #elif defined(__HAIKU__) || defined(__BEOS__)
     /* for intptr_t */
     #include <inttypes.h>
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/corestringlist.h browser/netsurf/utils/corestringlist.h
--- upstream/netsurf/utils/corestringlist.h	2026-09-20 23:29:00.183307243 +0100
+++ browser/netsurf/utils/corestringlist.h	2026-09-20 23:19:21.051169780 +0100
@@ -119,6 +119,16 @@
 CORESTRING_LWC_STRING(select);
 CORESTRING_LWC_STRING(src);
 CORESTRING_LWC_STRING(style);
+/* fixes195 - inline SVG corestrings (used by box_construct +
+ * macos9_svg_inline.c to identify <svg> roots and shape children). */
+CORESTRING_LWC_STRING(svg);
+CORESTRING_LWC_STRING(path);
+CORESTRING_LWC_STRING(ellipse);
+CORESTRING_LWC_STRING(polyline);
+CORESTRING_LWC_STRING(line);
+CORESTRING_LWC_STRING(defs);
+CORESTRING_LWC_STRING(g);
+CORESTRING_LWC_STRING(use);
 CORESTRING_LWC_STRING(submit);
 CORESTRING_LWC_STRING(table);
 CORESTRING_LWC_STRING(tbody);
@@ -248,6 +258,7 @@
 CORESTRING_DOM_STRING(hashchange);
 CORESTRING_DOM_STRING(height);
 CORESTRING_DOM_STRING(Home);
+CORESTRING_DOM_STRING(download);   /* fixes1063 (#114) */
 CORESTRING_DOM_STRING(href);
 CORESTRING_DOM_STRING(hreflang);
 CORESTRING_DOM_STRING(hspace);
diff -urN upstream/netsurf/utils/errors.h browser/netsurf/utils/errors.h
--- upstream/netsurf/utils/errors.h
+++ browser/netsurf/utils/errors.h
@@ -64,7 +64,7 @@
 	NSERROR_UNSAFE_REDIRECT,        /**< Fetch encountered a unsafe redirect */
 	NSERROR_BAD_AUTH,               /**< Fetch needs authentication data */
 	NSERROR_BAD_CERTS,              /**< Fetch needs certificate chain check */
-	NSERROR_TIMEOUT,                /**< Operation timed out */
+	NSERROR_TIMEOUT                 /**< Operation timed out */
 } nserror;
 
 #endif
diff -urN upstream/netsurf/utils/file.c browser/netsurf/utils/file.c
--- upstream/netsurf/utils/file.c
+++ browser/netsurf/utils/file.c
@@ -23,6 +23,7 @@
 #include <stdarg.h>
 #include <string.h>
 #include <stdio.h>
+#include <utils/dirent.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <unistd.h>
@@ -319,7 +320,7 @@
 nserror
 netsurf_recursive_rm(const char *path)
 {
-	DIR *parent;
+	MAC_DIR *parent;
 	struct dirent *entry;
 	nserror ret = NSERROR_OK;
 	struct stat ent_stat; /* stat result of leaf entry */
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/hashmap.c browser/netsurf/utils/hashmap.c
--- upstream/netsurf/utils/hashmap.c	2026-09-20 23:29:00.183307243 +0100
+++ browser/netsurf/utils/hashmap.c	2026-09-20 23:19:21.052077154 +0100
@@ -232,10 +232,12 @@
 bool
 hashmap_iterate(hashmap_t *hashmap, hashmap_iteration_cb_t cb, void *ctx)
 {
-	for (uint32_t bucket = 0;
+	uint32_t bucket;
+	hashmap_entry_t *entry;
+	for (bucket = 0;
 	     bucket < hashmap->bucket_count;
 	     bucket++) {
-		for (hashmap_entry_t *entry = hashmap->buckets[bucket];
+		for (entry = hashmap->buckets[bucket];
 		     entry != NULL;
 		     entry = entry->next) {
 			/* If the callback returns true, we early-exit */
diff -urN upstream/netsurf/utils/hashtable.c browser/netsurf/utils/hashtable.c
--- upstream/netsurf/utils/hashtable.c
+++ browser/netsurf/utils/hashtable.c
@@ -36,7 +36,8 @@
 #include <errno.h>
 
 #include "utils/log.h"
-#include "utils/hashtable.h"
+#include "utils/errors.h"
+#include "utils/hashtable.h"
 
 
 struct hash_entry {
@@ -128,8 +129,7 @@
 
 	res = hash_add(hash, (char *)key, (char *)value);
 	if (res != NSERROR_OK) {
-		NSLOG(netsurf, INFO,
-		      "Unable to add %s:%s to hash table", ln, value);
+		nslog_log(__FILE__, "", __LINE__, "Unable to add %s:%s to hash table", ln, value);
 	}
 	return res;
 }
@@ -196,7 +196,7 @@
 
 	ret = inflateInit2(&strm, 32 + MAX_WBITS);
 	if (ret != Z_OK) {
-		NSLOG(netsurf, INFO, "inflateInit returned %d", ret);
+		nslog_log(__FILE__, "", __LINE__, "inflateInit returned %d", ret);
 		return NSERROR_INVALID;
 	}
 
@@ -243,7 +243,7 @@
 	inflateEnd(&strm);
 
 	if (ret != Z_STREAM_END) {
-		NSLOG(netsurf, INFO, "inflate returned %d", ret);
+		nslog_log(__FILE__, "", __LINE__, "inflate returned %d", ret);
 		return NSERROR_INVALID;
 	}
 	return NSERROR_OK;
@@ -265,8 +265,9 @@
 	r->chain = calloc(chains, sizeof(struct hash_entry *));
 
 	if (r->chain == NULL) {
-		NSLOG(netsurf, INFO,
-		      "Not enough memory for %d hash table chains.", chains);
+		nslog_log(__FILE__, "", __LINE__,
+			      "Not enough memory for %d hash table chains.",
+			      chains);
 		free(r);
 		return NULL;
 	}
@@ -373,9 +374,10 @@
 
 	fp = gzopen(path, "r");
 	if (!fp) {
-		NSLOG(netsurf, INFO,
-		      "Unable to open file \"%.100s\": %s", path,
-		      strerror(errno));
+		nslog_log(__FILE__, "", __LINE__,
+			      "Unable to open file \"%.100s\": %s",
+			      path,
+			      strerror(errno));
 
 		return NSERROR_NOT_FOUND;
 	}
diff -urN upstream/netsurf/utils/hashtable.h browser/netsurf/utils/hashtable.h
--- upstream/netsurf/utils/hashtable.h
+++ browser/netsurf/utils/hashtable.h
@@ -25,6 +25,8 @@
 #define _NETSURF_UTILS_HASHTABLE_H_
 
 #include <stdbool.h>
+
+#include "utils/errors.h"
 
 struct hash_table;
 
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/inet.h browser/netsurf/utils/inet.h
--- upstream/netsurf/utils/inet.h	2026-09-20 23:29:00.184230646 +0100
+++ browser/netsurf/utils/inet.h	2026-09-20 23:19:21.056881481 +0100
@@ -49,6 +49,12 @@
 #define ns_close_socket CloseSocket
 #endif
 
+#elif defined(__MACOS9__)
+
+#include "frontends/macos9/shims/mac_inet.h"
+
+#define ns_close_socket(s) ((void)(s), 0)
+
 #else
 
 #include <winsock2.h>
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/log.c browser/netsurf/utils/log.c
--- upstream/netsurf/utils/log.c	2026-09-20 23:29:00.184230646 +0100
+++ browser/netsurf/utils/log.c	2026-09-20 23:19:21.057324659 +0100
@@ -266,18 +266,18 @@
 
 	/* sucessfull logging initialisation so log system info */
 	if (ret == NSERROR_OK) {
-		NSLOG(netsurf, INFO, "NetSurf version '%s'", netsurf_version);
+		nslog_log(__FILE__, "", __LINE__, "NetSurf version '%s'", netsurf_version);
 		if (uname(&utsname) < 0) {
 			NSLOG(netsurf, INFO,
 			      "Failed to extract machine information");
 		} else {
-			NSLOG(netsurf, INFO,
-			      "NetSurf on <%s>, node <%s>, release <%s>, version <%s>, machine <%s>",
-			      utsname.sysname,
-			      utsname.nodename,
-			      utsname.release,
-			      utsname.version,
-			      utsname.machine);
+			nslog_log(__FILE__, "", __LINE__,
+				      "NetSurf on <%s>, node <%s>, release <%s>, version <%s>, machine <%s>",
+				      utsname.sysname,
+				      utsname.nodename,
+				      utsname.release,
+				      utsname.version,
+				      utsname.machine);
 		}
 	}
 
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/messages.c browser/netsurf/utils/messages.c
--- upstream/netsurf/utils/messages.c	2026-09-20 23:29:00.184230646 +0100
+++ browser/netsurf/utils/messages.c	2026-09-20 23:19:21.057599692 +0100
@@ -37,7 +37,7 @@
 #include "utils/log.h"
 #include "utils/messages.h"
 #include "utils/utils.h"
-#include "utils/hashtable.h"
+#include "utils/hashtable.h"
 
 /** Messages are stored in a fixed-size hash table. */
 #define HASH_SIZE 101
@@ -126,9 +126,9 @@
 
 	nctx = messages_create_ctx(HASH_SIZE);
 	if (nctx == NULL) {
-		NSLOG(netsurf, INFO,
-		      "Unable to create hash table for messages file %s",
-		      path);
+		nslog_log(__FILE__, "", __LINE__,
+			      "Unable to create hash table for messages file %s",
+			      path);
 		return NSERROR_NOMEM;
 	}
 
@@ -180,7 +180,7 @@
 		return NSERROR_BAD_PARAMETER;
 	}
 
-	NSLOG(netsurf, INFO, "Loading Messages from '%s'", path);
+	nslog_log(__FILE__, "", __LINE__, "Loading Messages from '%s'", path);
 
 	return messages_load_ctx(path, &messages_hash);
 }
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/nsoption.c browser/netsurf/utils/nsoption.c
--- upstream/netsurf/utils/nsoption.c	2026-09-20 23:29:00.184230646 +0100
+++ browser/netsurf/utils/nsoption.c	2026-09-20 23:19:21.059192046 +0100
@@ -34,7 +34,7 @@
 
 #include "netsurf/inttypes.h"
 #include "netsurf/plot_style.h"
-#include "utils/errors.h"
+#include "utils/errors.h"
 #include "utils/log.h"
 #include "utils/utils.h"
 #include "utils/nsoption.h"
@@ -537,7 +537,7 @@
 
 	/* skip leading whitespace for start of key */
 	for (key = line; *key != 0; key++) {
-		if ((*key != ' ') && (*key != '\t') && (*key != '\n')) {
+		if ((*key != ' ') && (*key != '\t') && (*key != '\n') && (*key != '\r')) {
 			break;
 		}
 	}
@@ -567,7 +567,7 @@
 	}
 
 	/* remove delimiter from value */
-	if (line[linelen - 1] == '\n') {
+	while (linelen > 0 && (line[linelen - 1] == '\n' || line[linelen - 1] == '\r')) {
 		linelen--;
 		line[linelen] = 0;
 	}
@@ -721,11 +721,11 @@
 
 	fp = fopen(path, "r");
 	if (!fp) {
-		NSLOG(netsurf, INFO, "Failed to open file '%s'", path);
+		nslog_log(__FILE__, "", __LINE__, "Failed to open file '%s'", path);
 		return NSERROR_NOT_FOUND;
 	}
 
-	NSLOG(netsurf, INFO, "Successfully opened '%s' for Options file", path);
+	nslog_log(__FILE__, "", __LINE__, "Successfully opened '%s' for Options file", path);
 
 	while (fgets(s, NSOPTION_MAX_LINE_LEN, fp)) {
 		optionline(opts, s, strlen(s));
@@ -812,8 +812,7 @@
 
 	fp = fopen(path, "w");
 	if (!fp) {
-		NSLOG(netsurf, INFO, "failed to open file '%s' for writing",
-		      path);
+		nslog_log(__FILE__, "", __LINE__, "failed to open file '%s' for writing", path);
 		return NSERROR_NOT_FOUND;
 	}
 
@@ -902,7 +901,7 @@
 
 		/* arg+arglen is the option to set, val is the value */
 
-		NSLOG(netsurf, INFO, "%.*s = %s", arglen, arg, val);
+		nslog_log(__FILE__, "", __LINE__, "%.*s = %s", arglen, arg, val);
 
 		for (entry_loop = 0;
 		     entry_loop < NSOPTION_LISTEND;
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/string.h browser/netsurf/utils/string.h
--- upstream/netsurf/utils/string.h	2026-09-20 23:29:00.185965458 +0100
+++ browser/netsurf/utils/string.h	2026-09-20 23:19:21.061460574 +0100
@@ -24,10 +24,88 @@
 #ifndef _NETSURF_UTILS_STRING_H_
 #define _NETSURF_UTILS_STRING_H_
 
+/*
+ * On CW8, #include <string.h> finds THIS file instead of MSL's.
+ * Declare the standard C string functions that callers expect
+ * from <string.h> so they don't get implicit-int returns.
+ */
+#ifdef __MWERKS__
+#ifndef _MSL_STRING_H
+char *strchr(const char *s, int c);
+char *strrchr(const char *s, int c);
+char *strstr(const char *haystack, const char *needle);
+char *strdup(const char *s);
+char *strndup(const char *s, unsigned long n);
+int   strcmp(const char *s1, const char *s2);
+int   strncmp(const char *s1, const char *s2, unsigned long n);
+int   strcasecmp(const char *s1, const char *s2);
+int   strncasecmp(const char *s1, const char *s2, unsigned long n);
+char *strcpy(char *dst, const char *src);
+char *strncpy(char *dst, const char *src, unsigned long n);
+char *strcat(char *dst, const char *src);
+char *strncat(char *dst, const char *src, unsigned long n);
+unsigned long strlen(const char *s);
+void *memset(void *s, int c, unsigned long n);
+void *memcpy(void *dst, const void *src, unsigned long n);
+void *memmove(void *dst, const void *src, unsigned long n);
+int   memcmp(const void *s1, const void *s2, unsigned long n);
+char *strtok(char *str, const char *delim);
+char *strpbrk(const char *s, const char *accept);
+char *strsep(char **stringp, const char *delim);
+#endif
+#endif
+
 #include <stdlib.h>
 #include <stdarg.h>
 
-#include "utils/errors.h"
+/* CW8 cannot resolve includes from headers loaded via access path
+ * (any include style - bare or path-prefixed silently fails to open
+ * once we're inside this file). Inline the nserror enum directly,
+ * guarded by the same guard utils/errors.h uses, so a later real
+ * #include "utils/ns_errors.h" from a primary .c source short-circuits. */
+#ifndef NETSURF_UTILS_ERRORS_H_
+#define NETSURF_UTILS_ERRORS_H_
+typedef enum {
+	NSERROR_OK,
+	NSERROR_UNKNOWN,
+	NSERROR_NOMEM,
+	NSERROR_NO_FETCH_HANDLER,
+	NSERROR_NOT_FOUND,
+	NSERROR_NOT_DIRECTORY,
+	NSERROR_SAVE_FAILED,
+	NSERROR_CLONE_FAILED,
+	NSERROR_INIT_FAILED,
+	NSERROR_BMP_ERROR,
+	NSERROR_GIF_ERROR,
+	NSERROR_ICO_ERROR,
+	NSERROR_PNG_ERROR,
+	NSERROR_SPRITE_ERROR,
+	NSERROR_SVG_ERROR,
+	NSERROR_BAD_ENCODING,
+	NSERROR_NEED_DATA,
+	NSERROR_ENCODING_CHANGE,
+	NSERROR_BAD_PARAMETER,
+	NSERROR_INVALID,
+	NSERROR_BOX_CONVERT,
+	NSERROR_STOPPED,
+	NSERROR_DOM,
+	NSERROR_CSS,
+	NSERROR_CSS_BASE,
+	NSERROR_BAD_URL,
+	NSERROR_BAD_CONTENT,
+	NSERROR_FRAME_DEPTH,
+	NSERROR_PERMISSION,
+	NSERROR_NOSPACE,
+	NSERROR_BAD_SIZE,
+	NSERROR_NOT_IMPLEMENTED,
+	NSERROR_BAD_REDIRECT,
+	NSERROR_CYCLIC_REDIRECT,
+	NSERROR_UNSAFE_REDIRECT,
+	NSERROR_BAD_AUTH,
+	NSERROR_BAD_CERTS,
+	NSERROR_TIMEOUT
+} nserror;
+#endif
 
 
 /**
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/netsurf/utils/talloc.c browser/netsurf/utils/talloc.c
--- upstream/netsurf/utils/talloc.c	2026-09-20 23:29:00.185965458 +0100
+++ browser/netsurf/utils/talloc.c	2026-09-20 23:19:21.061731631 +0100
@@ -55,10 +55,16 @@
 #include "replace.h"
 #else
 #include <stdarg.h>
+/* MacSurf: include string.h and stdlib.h unconditionally rather than
+ * gating on __GNUC__ > 2. CW8 doesn't define __GNUC__ at all, so the
+ * upstream gate would skip these and rely on the prefix file pulling
+ * them in transitively. Explicit is safer. */
+#include <utils/string.h>
+#include <stdlib.h>
 #if !defined(__BEOS__) && __GNUC__ > 2
 /* Assume we've got va_copy */
 #define HAVE_VA_COPY
 #include <string.h>
 #endif
 #endif
 #include "talloc.h"
@@ -75,11 +80,27 @@
 #define TALLOC_FLAG_LOOP 0x02
 #define TALLOC_MAGIC_REFERENCE ((const char *)1)
 
-/* by default we abort when given a bad pointer (such as when talloc_free() is called 
+/* by default we abort when given a bad pointer (such as when talloc_free() is called
    on a pointer that came from malloc() */
 #ifndef TALLOC_ABORT
+#ifdef __MACOS9__
+/* fixes161f -- talloc abort goes to MSL's exit-cleanly path on Carbon, not
+ * MacsBug. Apple/huffpost crashes were diagnosed as silent process exits
+ * (log truncates, new session begins, MacsBug never fires). Log the reason
+ * before aborting so the next apple/huffpost reproduction tells us which
+ * site died (double-free vs corruption) and on what pointer. */
+#include "macsurf_debug.h"
+#define TALLOC_ABORT(reason) \
+	do { \
+		macsurf_debug_log_writef( \
+			"TALLOC_ABORT reason=%s ptr=%p", \
+			(reason), (void *)ptr); \
+		abort(); \
+	} while (0)
+#else
 #define TALLOC_ABORT(reason) abort()
 #endif
+#endif
 
 #ifndef discard_const_p
 #if defined(__intptr_t_defined) || defined(HAVE_INTPTR_T)
@@ -128,16 +149,60 @@
 #define TC_HDR_SIZE ((sizeof(struct talloc_chunk)+15)&~15)
 #define TC_PTR_FROM_CHUNK(tc) ((void *)(TC_HDR_SIZE + (char*)tc))
 
+/* fixes907 (reconvert double-free hunt) -- names the box-tree free currently in
+ * flight, set by html.c around the reconvert-old-tree free and the
+ * content-destroy bctx free. A TALLOC_ABORT then reports WHICH free path caught
+ * the corruption ("during=") plus WHAT chunk (talloc name/size), which localizes
+ * the double-free the hardware log shows on nav-away from a reconverted page. */
+const char *macsurf_talloc_free_ctx = "(none)";
+#ifndef __MACOS9__
+extern void macsurf_debug_log_writef(const char *fmt, ...);
+#endif
+
 /* panic if we get a bad magic value */
 static inline struct talloc_chunk *talloc_chunk_from_ptr(const void *ptr)
 {
 	const void *pp = ((const char *)ptr) - TC_HDR_SIZE;
 	struct talloc_chunk *tc = discard_const_p(struct talloc_chunk, pp);
-	if (unlikely((tc->flags & (TALLOC_FLAG_FREE | ~0xF)) != TALLOC_MAGIC)) { 
+	if (unlikely((tc->flags & (TALLOC_FLAG_FREE | ~0xF)) != TALLOC_MAGIC)) {
 		if (tc->flags & TALLOC_FLAG_FREE) {
-			TALLOC_ABORT("Bad talloc magic value - double free"); 
+			/* fixes907 -- freed chunk: header (incl. name) survives the
+			 * free, so tc->name is a live type/location literal.
+			 * fixes908 -- first-free= is the path that first freed this box
+			 * (from the box_talloc_destructor ring); during= is the path
+			 * doing the second (double) free right now. */
+			extern const char *macsurf_box_freed_lookup(void *box);
+			extern void *macos9_deathrow_cur_ptr;
+			extern void *macos9_deathrow_cur_fn;
+			/* fixes914 -- name the death-row entry being torn down, if
+			 * any. The 2026-07-19 abort fired inside a 32+ entry drain
+			 * and we could not say which entry; these two globals cost
+			 * nothing on the hot path and answer it. dr=(nil) means the
+			 * fault was NOT inside a death-row teardown. */
+			macsurf_debug_log_writef(
+				"TALLOC CORRUPT double-free chunk=%p size=%ld"
+				" name=%s during=%s first-free=%s dr=%p drfn=%p",
+				(void *)ptr, (long)tc->size,
+				tc->name ? tc->name : "?",
+				macsurf_talloc_free_ctx,
+				macsurf_box_freed_lookup((void *)ptr),
+				macos9_deathrow_cur_ptr, macos9_deathrow_cur_fn);
+			TALLOC_ABORT("Bad talloc magic value - double free");
 		} else {
-			TALLOC_ABORT("Bad talloc magic value - unknown value"); 
+			/* fixes907 -- magic wrong / never a chunk: name may be garbage,
+			 * so log only the numeric fields (no %s). */
+			macsurf_debug_log_writef(
+				/* fixes927 - %ld, not %lx: the hand-rolled
+				 * formatter has no hex and emits an unknown
+				 * specifier literally, so this line was
+				 * printing "flags=%lx" in the very corruption
+				 * report it exists for. Decimal is fine; the
+				 * value is compared against TALLOC_MAGIC. */
+				"TALLOC CORRUPT unknown chunk=%p flags=%ld size=%ld"
+				" during=%s",
+				(void *)ptr, (long)tc->flags,
+				(long)tc->size, macsurf_talloc_free_ctx);
+			TALLOC_ABORT("Bad talloc magic value - unknown value");
 		}
 	}
 	return tc;
@@ -1173,12 +1238,14 @@
 }
 
 #ifndef HAVE_VA_COPY
+#ifndef va_copy
 #ifdef HAVE___VA_COPY
 #define va_copy(dest, src) __va_copy(dest, src)
 #else
 #define va_copy(dest, src) (dest) = (src)
 #endif
 #endif
+#endif
 
 char *talloc_vasprintf(const void *t, const char *fmt, va_list ap)
 {	
diff -urN upstream/netsurf/utils/utf8.c browser/netsurf/utils/utf8.c
--- upstream/netsurf/utils/utf8.c
+++ browser/netsurf/utils/utf8.c
@@ -472,8 +472,8 @@
 
 	ret = guit->utf8->utf8_to_local(utf8_text, strlen(utf8_text), &conv);
 	if (ret != NSERROR_OK) {
-		NSLOG(netsurf, INFO,
-		      "failed to convert to local encoding, return %d", ret);
+		nslog_log(__FILE__, "", __LINE__,
+			 "failed to convert to local encoding, return %d", ret);
 		return false;
 	}
 
